Styles and classes (#272)
javascript-tutorial/it.javascript.info: 2-ui/1-document/08-styles-and-classes/article.md

# Stili e classi

Prima di cominciare a trattare i modi in cui JavaScript interagisce con gli stili e le classi, ecco una regola importante. Si spera che sia abbastanza scontata, ma non guasta ripeterla.

Ci sono in genere due modi di applicare uno stile ad un elemento:

1. Creare una classe CSS ed aggiungerla: `<div class="...">`
2. Scrivere direttamente le proprietà dentro `style`: `<div style="...">`.

JavaScript può modificare sia le classi che le proprietà all'interno di `style`.

Dovremmo sempre preferire le classi CSS a `style`. Quest'ultimo dovrebbe essere usato solo se le classi non sono sufficienti.

Per esempio, è ragionevole l'uso di `style` se calcoliamo dinamicamente le coordinate di un elemento e vogliamo impostarle con JavaScript, come in casi analoghi:

```js
let top = /* calcoli complessi */;
let left = /* calcoli complessi */;

elem.style.left = left; // es '123px', calcolato al momento dell'esecuzione
elem.style.top = top; // es '456px'
```

> Per gli altri casi, come rendere rosso un testo o aggiungere un'icona di sfondo, definiamo degli stili CSS e poi applichiamo la classe con JavaScript. È più flessibile e più facile da gestire.

Author: Avevo pensato la stessa cosa. Comunque anche se non fosse un errore e l'autore avesse inteso proprio icona, un'icona è, di fatto, un'immagine. Pienamente d'accordo!

</body>
```

```smart header="Valori computed e resolved"
Ci sono due concetti nei [CSS](https://drafts.csswg.org/cssom/#resolved-values):

1. Un valore di uno stile *computed* è il valore derivante dall'applicazione di tutte le regole e dell'ereditarietà dei CSS: il risultato della cascata CSS. Può essere simile a `height:1em` o `font-size:125%`.
2. Un valore di uno stile *resolved* è quello applicato realmente sull'elemento. Valori come `1em` o `125%` sono relativi. Il browser prende i valori computed e li converte tutti in unità di misura fisse e assolute, per esempio: `height:20px` o `font-size:16px`. Per le proprietà geometriche i valori resolved possono avere la virgola, come `width:50.5px`.

Molto tempo addietro `getComputedStyle` è stato creato per ottenere i valori computed, ma si scoprì che i valori resolved erano molto più utili e lo standard è cambiato.

> Attualmente, quindi, restituisce il valore resolved della proprietà, di solito espressa in `px` per le proprietà geometriche.

Reviewer: ```suggestion
Attualmente, quindi, `getComputedStyle` restituisce il valore resolved della proprietà, di solito espressa in `px` per le proprietà geometriche.
```

```

````warn header="`getComputedStyle` richiede il nome completo della proprietà"
Dovremmo sempre chiedere l'esatta proprietà che si desidera, come `paddingLeft` o `marginTop` o `borderTopWidth`. Diversamente il risultato corretto non è garantito.

> Per esempio, se ci sono le proprietà `paddingLeft/paddingTop`, cosa dovremmo ottenere con `getComputedStyle(elem).padding`? Niente, o forse un valore "generato" a partire dai valori di padding noti? Non esiste una regola standard in questo caso.

Reviewer: ```suggestion
Per esempio, se ci sono le proprietà `paddingLeft/paddingTop`, cosa otterremmo con `getComputedStyle(elem).padding`? Niente, o forse un valore "generato" a partire dai valori di padding noti? Non esiste una regola standard in questo caso.
```

</style>
<script>
  let style = getComputedStyle(document.body);
  alert(style.margin); // Firefox restituisce una stringa vuota
</script>
```
````

```smart header="Gli stili applicati ai link `:visited` sono nascosti!"
I link visitati possono ricevere un colore tramite la pseudoclasse CSS `:visited`.

> Tuttavia `getComputedStyle non dà accesso a quel colore, poiché diversamente una pagina arbitraria potrebbe scoprire se l'utente ha visitato un link creandolo sulla pagina e verificandone gli stili.

Reviewer: ```suggestion
Tuttavia `getComputedStyle non dà accesso a quel colore, poiché diversamente, una pagina arbitraria potrebbe scoprire se l'utente ha visitato un link creandolo e verificandone gli stili.
```

JavaScript potrebbe non rilevare gli stili applicati da `:visited`. C'è una limitazione nei CSS, inoltre, che vieta su `:visited` l'applicazione di stili che modifichino le proprietà geometriche. Questo per garantire che, per una pagina malevola, non ci sia una via traversa di testare se un link è stato visitato e quindi di violare la privacy.
```

## Riepilogo

> Per gestire le classi ci sono due proprietà DOM:

Reviewer 2: ```suggestion
Per gestire le classi abbiamo a disposizione due proprietà DOM:
```

- `className` -- il valore stringa, ottimo per gestire l'intero insieme delle classi.

> - `classList` -- il valore oggetto con i metodi `add/remove/toggle/contains`, ottimo per gestire le classi individualmente.

Reviewer: ```suggestion
- `classList` -- l'oggetto con i metodi `add/remove/toggle/contains`, ottimo per gestire le classi individualmente.
```
